lib/wlib/strings/DynamicString.cpp

```diff
@@ -0,0 +1,482 @@
+     * Compares a string and character and return 0 if they are equal, less than 0 if
+     * given string is less than current string and greater than 0 if
+     * given string is greater than current string
+     *
+     * @param c character to compare against current string
+     * @return a signed number based on how strings compare
+     */
+    diff_type DynamicString::compare(const char c) const {
+        char array[2] = {c, '\0'};
+        return (diff_type) strcmp(this->c_str(), array);
+    }
+
+    /**
+    * Comparative operator compares if two strings are equal or not
+    *
+    * @param lhs @code DynamicString string as left hand side string
+    * @param rhs @code DynamicString string as right hand side string
+    * @return true or false based on if two strings are equal
+    */
+
+    bool operator==(const DynamicString &lhs, const DynamicString &rhs) {
+        return lhs.compare(rhs) == 0;
+    }
+
+    /**
+    * Comparative operator compares if two strings are equal or not
+    *
+    * @param lhs character string as left hand side string
+    * @param rhs @code DynamicString @endcode string as right hand side string
+    * @return true or false based on if two strings are equal
+    */
+    bool operator==(const char *lhs, const DynamicString &rhs) {
+        return rhs.compare(lhs) == 0;
+    }
+
+    /**
+    * Comparative operator compares if two strings are equal or not
+    *
+    * @param lhs @code DynamicString @endcode string as left hand side string
+    * @param rhs character string as right hand side string
+    * @return true or false based on if two strings are equal
+    */
+    bool operator==(const DynamicString &lhs, const char *rhs) {
+        return lhs.compare(rhs) == 0;
+    }
+
+    /**
+    * Comparative operator compares if string and single character are equal or not
+    *
+    * @param lhs character as left hand side
+    * @param rhs @code DynamicString string as right hand side string
+    * @return true or false based on if two strings are equal
+    */
+    bool operator==(const char lhs, const DynamicString &rhs) {
+        return rhs.compare(lhs) == 0;
+    }
+
+    /**
+    * Comparative operator compares if string and single character are equal or not
+    *
+    * @param lhs @code DynamicString string as left hand side string
+    * @param rhs character as right hand side
+    * @return true or false based on if two strings are equal
+    */
+    bool operator==(const DynamicString &lhs, const char rhs) {
+        return lhs.compare(rhs) == 0;
+    }
+
+    /**
+    * Additive operator adds two given strings
+    *
+    * @param lhs @code DynamicString string as left hand side string
+    * @param rhs @code DynamicString string as right hand side string
+    * @return a @code DynamicString that is build from left hand string and right hand string
+    */
+    DynamicString operator+(const DynamicString &lhs, const DynamicString &rhs) {
+        DynamicString newStr(lhs.c_str());

```

Comment:
Better than before, but this does an allocation for `lhs.length()` and copies `lhs`, then allocates `lhs.length() + rhs.length()` and then copies `lhs` and `rhs`. You should add a constructor `DynamicString(const char *str1, const char *str2, size_type len1, size_type len2)` which allocates `len1 + len2` and then copies `str1` and `str2`. You could then call `return DynamicString(lhs.c_str(), rhs.c_str(), lhs.length(), rhs.length())` and equivalent code for regular C Strings